src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +249,140 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+                catch (Exception e)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    return null;
+                }
+            }
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Sends CLNT_UCAST_INST request for given instance name to SQL Sever Browser, and receive SVR_RESP from the Browser.
+        /// </summary>
+        /// <param name="browserEP">SQL Sever Browser IPEndPoint</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <returns>SVR_RESP packets from SQL Sever Browser</returns>
+        private static byte[] SendInstanceNameRequest(string browserHostname, string instanceName)
+        {
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
+            }
+
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = 0x04;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
+
+            byte[] responsePacket = null;
+            UdpClient client = null;
+            try
+            {
+                client = new UdpClient(AddressFamily.InterNetwork);
+                Task<int> sendTask = client.SendAsync(requestPacket, requestPacket.Length, browserHostname, SQL_SVR_BROWSER_PORT);
+                Task<UdpReceiveResult> receiveTask = null;
+                if (sendTask.Wait(1000) && (receiveTask = client.ReceiveAsync()).Wait(1000))
+                {
+                    responsePacket = receiveTask.Result.Buffer;
+                }
+            }
+            finally
+            {
+                if (client != null)
+                {
+                    client.Dispose();
+                }
+            }
+
+            return responsePacket;
+        }
 
-            return new SNITCPHandle(serverAndPortParts[0], portNumber, timerExpire, callbackObject, parallel);
+        /// <summary>
+        /// Finds port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <returns>port number for given instance name</returns>
+        private static int GetPortByInstanceName(string browserHostname, string instanceName)
+        {
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
```

Comment:
I don't think this `ArgumentException` be thrown at the caller or should surface to any Calling functions. 
Debug assert may be a better way to take care of these scenario. 

What inputs do you expect would lead to an `ArgumentException` in this case? 
For the same input, what is the existing behavior on Windows?